### PR TITLE
ISPN-14211 Add Keycloak no-downtime index changes scenarios to Infini…

### PR DESCRIPTION
**Pull request by mhajas**
…span testsuite

This PR adds test scenarios for changing (adding/removing/migrating from non-analyzed to analyzed and vice versa) Inifnispan indexes in no downtime way.

This is my first PR to this repository. Please let me know if I miss something necessary for merging this.

### CODE REVIEW

**Review by fax4ever — CHANGES_REQUESTED** [review]
Thank you Michal for the nice job.
It is great to have these exhaustive test cases on Infinispan code base.
I would suggest some changes here.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/evolution/model/ModelD.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    @ProtoField(number = 2)
+    public String id;
+
+    @ProtoField(number = 3)
+    @Deprecated
+    public String name;
+
+    @ProtoField(number = 4)
+    @ProtoDoc("@Field(index = Index.YES, store = Store.YES, analyze = Analyze.YES, analyzer = @Analyzer(definition = \"lowercase\"))")
```
[the remark this answers is not in the document]
Comment (mhajas) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/evolution/model/ModelD.java:24]: @fax4ever I don't think removing `D` and `E` as it is used in some migration path as is. The thing that is probably causing confusion here, is this:
```
@ProtoDoc("@Field(index = Index.YES, store = Store.YES, analyze = Analyze.YES, analyzer = @Analyzer(definition = \"lowercase\"))")
```

is supposed to represent an analyzed field, while `@Keyword()` results in non-analyzed area (I hope I am not mistaken).

**Comment by mhajas** [comment]
@fax4ever thank you very much for your review. I added an updated version. I decided to change the codebase a bit as I think the usage of ModelA-J was really confusing, so I tried to express the purpose of each entity within its class name. I hope I didn't cause even more confusion.

While I was changing this, I realized that maybe we should also consider the migration paths for projectable/sortable/normalizer fields, WDYT? 

Some questions:
Is it possible to easily migrate non-sortable to sortable field? Is there any performance penalty if we have all indexed fields sortable and projectable even though we don't necessarily need this at the moment, but we may need it in the future?

**Comment by fax4ever** [comment]
Thank you @mhajas. Changes look good.
I added a small commit just to change some formatting.
Going to merge it as soon as the CI looks good.

**Comment by fax4ever** [comment]
> While I was changing this, I realized that maybe we should also consider the migration paths for projectable/sortable/normalizer fields, WDYT?
> 
> Some questions: Is it possible to easily migrate non-sortable to sortable field? Is there any performance penalty if we have all indexed fields sortable and projectable even though we don't necessarily need this at the moment, but we may need it in the future?

With the `RemoteCacheManagerAdmin#updateIndexSchema` the pre existing index data is preserved, so the indexes are not recreated from the existing data as in case of `#reindexCache`.

In case of migration of a given field from `non-sortable` to `sortable` for instance I think that it would be better to reindex the data, so that the Lucene extra data to address the sortability of the field would be created also for the preexing entities.

The same reason I think is valid for projectability. If a field is projectable, an extra Lucene field to store the data is created. Thus if we change the schema making the field projectable, the old entries would not have these extra data when we migrate the schema; and probably we would get a runtime exception if try to project the field from the old entries.

On the other hand, making a sortable (projectable) field non-sortable (non-projectable) I think that is safe.

About the idea of adding or changing an analyzer / normalizer, I think that the reindex is required to make the inverted indexes for the old (preexisting) entities consistent with the new analysis definition.

I don't know if @yrodiere can confirm it.

**Comment by yrodiere** [comment]
> I don't know if @yrodiere can confirm it.

Capabilities such as sortable, searchable, projectable, etc. need to be declared in the schema precisely because they are index-time things: we need to know that before we index, and the content of the index will be different based on that.

Same with analyzers/normalizers.

So yes, changing any of that requires reindexing. If you don't, at best older documents will be ignored, at worst you could get exceptions.

On that note, I'm currently on leave, good luck :)

**Comment by fax4ever** [comment]
> > I don't know if @yrodiere can confirm it.
> 
> Capabilities such as sortable, searchable, projectable, etc. need to be declared in the schema precisely because they are index-time things: we need to know that before we index, and the content of the index will be different based on that.
> 
> Same with analyzers/normalizers.
> 
> So yes, changing any of that requires reindexing. If you don't, at best older documents will be ignored, at worst you could get exceptions.
> 
> On that note, I'm currently on leave, good luck :)

@yrodiere thanks for the feedback and sorry to have bothered you. take care!

**Comment by fax4ever** [comment]
@pruivo just rebased on top of the current main

**Comment by fax4ever** [comment]
now CI looks good! Then going to merge it

**Comment by fax4ever** [comment]
thanks @mhajas for the contribution!

**Pull request merged**